Sources/Attention.metal

```diff
@@ -19,15 +19,16 @@ constant uint R [[function_constant(0)]];
 constant uint C [[function_constant(1)]];
 constant uint H [[function_constant(2)]];
 constant uint D [[function_constant(3)]];
+constant uint H_k [[function_constant(4)]];
 
 // Whether each matrix is transposed.
 constant bool Q_trans [[function_constant(10)]];
 constant bool K_trans [[function_constant(11)]];
 constant bool V_trans [[function_constant(12)]];
 constant bool O_trans [[function_constant(13)]];
 constant uint Q_leading_dim = Q_trans ? R : H * D;
-constant uint K_leading_dim = K_trans ? H * D : C;
-constant uint V_leading_dim = V_trans ? C : H * D;
+constant uint K_leading_dim = K_trans ? H_k * D : C;
```

Comment:
```suggestion
constant uint K_leading_dim = K_trans ? (grouped_query ? H_k : H) * D : C;
```

This is a breaking ABI/API change. The user now needs to specify 5 constants for a compute PSO to compile, compared to 4 before then change. A better solution for constant folding might fall back to `H` if there is no grouped query. Do all flavors of grouped query require two `H` variables?

Is it possible to compute `H_k` from other variables specified through compiler constants? For example, if it happens to equal `H / batch_size`, we would encode that expression instead of introducing a new constant. That is a general design rule of MFA, defer as much as possible to codegen.

I would suggest introducing a new variable, preferably not having a name collision of `H`, that can generate `H_k` as a function of `H`. If someone can create an analytical function to compute `H_k` in such a way. Makes it more expressive (you can understand what `H_k` means because you see the expression for it). Otherwise, I suggest entering a name in all lowercase, which I'll have to review for careful API design. Although function constants are specifiable by number, people can technically access them by name. We can't go back on a name after it's part of a stable ABI/API.

It's okay if we translocate the position where `Q/K/V/O_leading_dim` are declared in source code, if some function constants they depend on appear later in the function constant indices now. That affect may ripple down to more of the function constants, requiring a large amount of code to be shifted downward several lines. Doesn't change the ABI or API, just makes the source code potentially a bit less legible.

Reply:
> H_k is also smaller than H and H % H_k == 0.

If the integer modulus is always zero, that means `H_k` can be computed using an integer division over `H`. That would eliminate the `H_k` parameter and replace it with an expression `H / something`. In constant folding, we could specify `1` for that something unless an new function constant returns true for `is_function_constant_defined`.

```metal
constant uint some_name1 [[function_constant(some_index)]];
constant uint some_name2 = is_function_constant_defined(some_name1) ? some_name1 : 1;
constant uint K_leading_dim = K_trans ? (H / some_name2) * D : C;
```

Reply:
The ABI brokage is a good point. Yes, it is commonly referred to as `h_h_k_ratio` in Flash Attention's repo. I can look into the is_function_constant_defined thing. Otherwise do you want to keep the query_offsets or do you want to hard-code it into `gid.y / h_h_k_ratio`? (I find it useful to some degree, for example, I previously mistook this as `gid.y % H_k` and it is easy to fix outside of the shader code.

Reply:
If it's possible to execute this with query offsets, and without any function constants, that should already be how it's implemented. I assume it's impossible with the current ABI, which motivated the introduction of `H_k` into shader code.

Is `h_h_k_ratio` how it's spelled verbatim in the main FlashAttention repo? Including case of characters? If so, that would be the perfect name for the function constant. All lower-case.

Reply:
I think the matrix offsets or head offsets approach is more general. Right now, the popular thing is to make the queries a divisor of the total head count. But in the future, there might be other variants that aren't perfectly divisible. We want to be as general-purpose as possible. There also isn't very much overhead to fetching the query offset from RAM.

Reply:
Ha, just saw the comment on lower case. Yes, it is used that way, per: https://github.com/Dao-AILab/flash-attention/blob/main/csrc/flash_attn/src/flash.h#L46

Reply:
It does seem like this is necessary, and cannot simply be implemented through query offsets. We have pointer arithmetic inside the inner loop that depends on `h_h_k_ratio` if certain matrices are transposed. This arithmetic must be optimized away during AIR codegen to maximize performance.

<img width="822" alt="Screenshot 2024-01-04 at 12 50 53 PM" src="https://github.com/philipturner/metal-flash-attention/assets/71743241/84ee5685-eb86-4404-b610-32169c4177a5">

We might as well remove query offsets from the ABI, as nobody is using them. They are an attempted method to implement grouped query, which proved inappropriate. Now they are a software maintenance burden and technical debt.

To match the lower/uppercase convention used in MFA, I prefer `H_H_K_ratio` and `H_K`. The function constant index of 4 is appropriate, as this number is so vital to the pointer arithmetic. `is_function_constant_defined` is also needed because making this new constant mandatory would surely break existing clients.

Reply:
Yeah, this is necessary change. I also tested both in ccv / MFA to make sure it is not ABI-breaking / works for grouped query.

Reply:
Yeah, LLaMA 2 should be the 70B variant that uses GQA. The `query_offsets` saved me one time because I read it wrong and thought it is `y % H_k` rather than `y / H_H_k_ratio`. It is easy to change with `query_offsets` without recompiling the kernel.

I also need to read more about so-called single query attention, and whether that is a variant of GQA or something else entirely.

I think @tridao 's implementation right now only support the H_H_k_ratio based GQA, I could be wrong though.

Reply:
> It is easy to change with `query_offsets` without recompiling the kernel.

That is a strong indication we should keep query offsets. But if we keep query offsets, the function constants for dividing H by other numbers should be more flexible. Perhaps something like Q_stride, O_stride, K_stride, V_stride, etc. Reserve a namespace in the function constants with room for all those letters. Only populate it with the value for `H` for the time being. It will require some time to think about what is the most appropriate way to go about this.